src/PointwiseFunctions/Xcts/SpacetimeQuantities.cpp

```diff
@@ -113,137 +99,143 @@ void SpacetimeQuantitiesComputer::operator()(
-  for (size_t i = 0; i < 3; ++i) {
-    for (size_t j = 0; j < 3; ++j) {
-      for (size_t k = 0; k <= j; ++k) {
-        for (size_t l = 0; l < 3; ++l) {
-          deriv_extrinsic_curvature->get(i, j, k) -=
-              christoffel.get(l, i, j) * extrinsic_curvature.get(l, k);
-          deriv_extrinsic_curvature->get(i, j, k) -=
-              christoffel.get(l, i, k) * extrinsic_curvature.get(j, l);
-        }
-      }
-    }
-  }
+    detail::LongitudinalShiftMinusDtConformalMetric<DataVector> /*meta*/)
+    const {
+  const auto& longitudinal_shift_excess = cache->get_var(
+      *this, Tags::LongitudinalShiftExcess<DataVector, 3, Frame::Inertial>{});
+  TensorExpressions::evaluate<ti_I, ti_J>(
+      longitudinal_shift_minus_dt_conformal_metric,
+      longitudinal_shift_excess(ti_I, ti_J) +
+          longitudinal_shift_background_minus_dt_conformal_metric(ti_I, ti_J));
+}
+
+void SpacetimeQuantitiesComputer::operator()(
+    const gsl::not_null<tnsr::ii<DataVector, 3>*> extrinsic_curvature,
+    const gsl::not_null<Cache*> cache,
+    gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector> /*meta*/)
+    const {
+  const auto& lapse = cache->get_var(*this, gr::Tags::Lapse<DataVector>{});
+  const auto& longitudinal_shift_minus_dt_conformal_metric = cache->get_var(
+      *this, detail::LongitudinalShiftMinusDtConformalMetric<DataVector>{});
+  Xcts::extrinsic_curvature(
+      extrinsic_curvature, conformal_factor, lapse, conformal_metric,
+      longitudinal_shift_minus_dt_conformal_metric, trace_extrinsic_curvature);
 }
 
 void SpacetimeQuantitiesComputer::operator()(
     const gsl::not_null<Scalar<DataVector>*> hamiltonian_constraint,
     const gsl::not_null<Cache*> cache,
     gr::Tags::HamiltonianConstraint<DataVector> /*meta*/) const {
-  const auto& ricci_tensor = cache->get_var(
-      *this, gr::Tags::SpatialRicci<3, Frame::Inertial, DataVector>{});
-  const auto& extrinsic_curvature = cache->get_var(
-      *this, gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector>{});
+  const auto& conformal_laplacian_of_conformal_factor = cache->get_var(
+      *this, detail::ConformalLaplacianOfConformalFactor<DataVector>{});
   const auto& inv_spatial_metric = cache->get_var(
       *this, gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>{});
-  const auto& extrinsic_curvature_square =
-      cache->get_var(*this, detail::ExtrinsicCurvatureSquare<DataVector>{});
-  get(*hamiltonian_constraint) =
-      get(trace(ricci_tensor, inv_spatial_metric)) +
-             square(get(trace(extrinsic_curvature, inv_spatial_metric))) -
-             get(extrinsic_curvature_square);
+  const auto& extrinsic_curvature = cache->get_var(
+      *this, gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector>{});
+  // Eq. 3.12 in BaumgarteShapiro, divided by 2 for consistency with SpEC
+  TensorExpressions::evaluate(
+      hamiltonian_constraint,
+      4. * conformal_laplacian_of_conformal_factor() -
```

Comment:
I'm not sure if this will matter, but it might be (slightly) faster to factor out the factor of 0.5 and multiply by it after. This (I think) would reduce the number of FLOPs by ~3 x num_points. Totally up to you!

Reply:
Sure why not, added a few parens